Bluetooth: Host: deprecate CONFIG_BT_FIXED_PASSKEY

HaavardRei opened the pull request:
Deprecates the CONFIG_BT_FIXED_PASSKEY config and adds CONFIG_BT_APP_PASSKEY as an alternative.

Updates the bt shell implementation to use the new config option.

m-alperen-sener commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -726,12 +726,26 @@ config BT_SMP_USB_HCI_CTLR_WORKAROUND
+	  being cleared. The user can monitor that the passkey has been used through the
+	  @ref bt_conn_auth_info_cb.pairing_complete and @ref bt_conn_auth_info_cb.pairing_failed
+	  callbacks (the key will be cleared whether or not the pairing is successful). If set, the
+	  pairing_confirm() callback will be called for all incoming pairings.
```
```suggestion
	  pairing_confirm() callback will be called once for each passkey set.
```

m-alperen-sener commented on subsys/bluetooth/host/smp.c:
```diff
@@ -4430,9 +4442,14 @@ __maybe_unused static uint8_t display_passkey(struct bt_smp *smp)
 		smp->passkey = fixed_passkey;
+
+		if (IS_ENABLED(CONFIG_BT_ONE_TIME_PASSKEY)) {
+			fixed_passkey = BT_PASSKEY_INVALID;
```
here also we can set in_use

m-alperen-sener reviewed (changes requested):
Looks good small comments and suggestion.
Should we update the API documentation for bt_passkey_set() in conn.h ?

alwa-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -732,6 +732,18 @@ config BT_FIXED_PASSKEY
 	  pairing_confirm() callback will be called for all incoming pairings.
 
+config BT_ONE_TIME_PASSKEY
+	bool "Use a user-provided passkey for pairing"
```
`depends on !BT_FIXED_PASSKEY`?

HaavardRei committed:
Bluetooth: Host: Add BT_APP_PASSKEY Kconfig option

Adds the BT_APP_PASSKEY Kconfig, which allows the application to provide
passkeys for pairing using the new `app_passkey()` callback.
This is an alternative to BT_FIXED_PASSKEY, which will be deprecated in
a later commit.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
Bluetooth: Host: Deprecate BT_FIXED_PASSKEY

The BT_FIXED_PASSKEY Kconfig option is being deprecated, and is replaced
by BT_APP_PASSKEY. The reason for the deprecation is an upcoming
errata, ES-24489, which mandates that a new passkey shall be generated
for each pairing procedure.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
Bluetooth: Host: shell: Don't use BT_FIXED_PASSKEY

Updates the bt shell API to use the new Kconfig option
BT_APP_PASSKEY instead of BT_FIXED_PASSKEY as this is being
deprecated.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

m-alperen-sener commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -2663,6 +2672,23 @@ struct bt_conn_auth_cb {
+	/** @brief Allow the application to provide a passkey for pairing.
+	 *
+	 *  If implemented, this callback allows the application to provide passkeys for pairing.
+	 *  The valid range of passkeys is 0 - 999999, and should be expected to be zero-padded to
```
How do we do that? we return uint32_t?

PavelVPV commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2504,10 +2510,24 @@ static uint8_t legacy_request_tk(struct bt_smp *smp)
+
+			if (passkey != BT_PASSKEY_RAND) {
+				if (passkey > 999999) {
+					return BT_SMP_ERR_INVALID_PARAMS;
```
This should probably be `BT_SMP_ERR_UNSPECIFIED` because `BT_SMP_ERR_INVALID_PARAMS` means `The Invalid Parameters error code indicates that the command
length is invalid or that a parameter is outside of the specified range` and belongs to SMP. The error code returned here will be sent to peer device and peer may treat it incorrectly.

I think it worth to add a warning here so that application knows what happened.

PavelVPV reviewed (commented):
LGTM, 2 minor comments.

Also, it would probably be good to have some simple multilink bsim test to check parallel pairing, but this is up to you. We can do this separately later as well.

HaavardRei commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2504,10 +2510,24 @@ static uint8_t legacy_request_tk(struct bt_smp *smp)
+		}
+
+#if defined(CONFIG_BT_APP_PASSKEY)
+		if (smp_auth_cb && smp_auth_cb->app_passkey) {
```
Tested legacy pairing locally without any issues!

m-alperen-sener commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -725,12 +725,24 @@ config BT_SMP_USB_HCI_CTLR_WORKAROUND
+	help
+	  With this option enabled, the application will be able to provide passkeys for pairing
+	  using the app_passkey() callback.  If the application does not provide a passkey, a
+	  random passkey will be generated.
```
```suggestion
	  random passkey will be generated by the Host.
```

m-alperen-sener reviewed (approved):
I left the last comment to make it a bit more clear for user that host will keep generating random value. Up to you, just suggestions.

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -2663,6 +2673,23 @@ struct bt_conn_auth_cb {
+	 *  If implemented, this callback allows the application to provide passkeys for pairing.
+	 *  The valid range of passkeys is 0 - 999999. The application shall return the passkey for
+	 *  pairing, or BT_PASSKEY_RAND to generate a random passkey.
+	 *
```
The presence (not-NULL-ness) of this callback has an effect. We need to document that.

HaavardRei replied:
I'm not sure I understand what you mean. Do you mean that the application is responsible for pairing keys when this is implemented? If so, the note below mentions this. Otherwise, please add a suggestion!

alwa-nordic replied:
Resolved. Effect is technically documented. We agreed accessibility of documentation does not need to be prioritized for this API, so we don't need to give further guidance in the documentation.

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -2663,6 +2673,23 @@ struct bt_conn_auth_cb {
+	 *  If implemented, this callback allows the application to provide passkeys for pairing.
+	 *  The valid range of passkeys is 0 - 999999. The application shall return the passkey for
+	 *  pairing, or BT_PASSKEY_RAND to generate a random passkey.
+	 *
```
Should we specify that blocking in this callback is not allowed? I think we should.

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -2663,6 +2673,23 @@ struct bt_conn_auth_cb {
+	 *  If implemented, this callback allows the application to provide passkeys for pairing.
+	 *  The valid range of passkeys is 0 - 999999. The application shall return the passkey for
+	 *  pairing, or BT_PASSKEY_RAND to generate a random passkey.
+	 *
```
I mean it only works when (1) Legacy Pairing occurs and (2) the Core v6.1 3.H.2.3.5.1 Table 2.8 determines the local device to be the display.

About (1), I skimmed `smp.c` now and it does not look like Legacy Pairing is forced when this callback or a fixed passkey is set. The user probably want to disable LE Secure Connections when they want their app to provide the passkey.

As for (2), the right description depends on what we decide in https://github.com/zephyrproject-rtos/zephyr/pull/97597#discussion_r2445084013.

HaavardRei commented:
Resolved comments from @alwa-nordic offline, but changed the IO capabilities when using the app_passkey cb to support both BT_SMP_IO_KEYBOARD_DISPLAY (when defining the passkey_entry cb and app_passkey) and BT_SMP_IO_DISPLAY_ONLY (when defining just app_passkey)
